src/ui/components/pages/importEmail/signContext.tsx

```diff
@@ -0,0 +1,161 @@
+
+import { Login } from './login';
+import * as styles from './signWrapper.module.css';
+
+export const SignContext = createContext<{
+  openConfirmModal: () => Promise<unknown>;
+  closeConfirmModal: () => void;
+}>({
+  openConfirmModal: async () => undefined,
+  closeConfirmModal: () => undefined,
+});
+
+export const SignProvider = ({ children }: { children: ReactNode }) => {
+  const { t } = useTranslation();
+
+  const account = usePopupSelector(state => state.selectedAccount);
+
+  const [modalOpen, setModalOpen] = useState(false);
+
+  const modalPromiseRef = useRef<{
+    resolve: null | ((value: unknown) => void);
+  }>({ resolve: null });
+
+  const openConfirmModal = useCallback(() => {
+    setModalOpen(true);
+
+    return new Promise(resolve => {
+      modalPromiseRef.current = { resolve };
+    });
+  }, []);
+
+  const closeConfirmModal = useCallback(() => {
+    setModalOpen(false);
+  }, []);
+
+  const onOk = useCallback(() => {
+    modalPromiseRef.current.resolve?.(true);
+    modalPromiseRef.current.resolve = null;
+  }, []);
+
+  const onCancel = useCallback(() => {
+    setModalOpen(false);
+    modalPromiseRef.current.resolve?.(false);
+    modalPromiseRef.current.resolve = null;
+  }, []);
+
+  const contextValue = useMemo(
+    () => ({ openConfirmModal, closeConfirmModal }),
+    [closeConfirmModal, openConfirmModal]
+  );
+
+  return (
+    <>
+      <SignContext.Provider value={contextValue}>
+        {children}
+      </SignContext.Provider>
+
+      {account?.type === 'wx' && (
+        <Modal showModal={modalOpen} animation={Modal.ANIMATION.FLASH}>
+          <div className={clsx('modal', 'cover', styles.root)}>
+            <div className={styles.content}>
+              <Button
+                className="modal-close"
+                onClick={onCancel}
+                type="button"
+                view="transparent"
+              />
+
+              <h2 className={clsx('margin4', 'title1')}>
+                {t('importEmail.loginRequired')}
+              </h2>
+
+              <Login
+                userData={{ username: account.username, password: '' }}
+                onConfirm={onOk}
+              />
+            </div>
+          </div>
+        </Modal>
+      )}
+
+      {account?.type === 'ledger' && (
+        <Modal animation={Modal.ANIMATION.FLASH} showModal={modalOpen}>
+          <LedgerConnectModal
+            networkCode={account.networkCode}
+            onClose={onCancel}
+            onReady={onOk}
+          />
+        </Modal>
+      )}
+    </>
+  );
+};
+
+export function useSign<OnConfirmParams>(
+  onConfirm: (params: OnConfirmParams) => void | Promise<void>
+): [(params: OnConfirmParams) => Promise<void>, boolean] {
+  const account = usePopupSelector(state => state.selectedAccount);
+
+  const [pending, setPending] = useState(false);
+
+  const { openConfirmModal, closeConfirmModal } = useContext(SignContext);
+
+  const sign = useCallback(
```

Comment:
Can this function be passed from provider instead, so all this logic for wallet types will be closer to each other? And if it's possible to get rid of ref keeping resolve function in provider that would be really great, because it's sometimes hard to follow the code with these kinds of things involved